pkg/controllers/provisioning/scheduling/nodeset.go

```diff
@@ -76,3 +102,583 @@ func (s *NodeSet) getDaemons(ctx context.Context, constraints *v1alpha5.Constrai
 func (s *NodeSet) Add(node *Node) {
 	s.nodes = append(s.nodes, node)
 }
+
+func (s *NodeSet) SchedulePod(ctx context.Context, pod *v1.Pod) error {
+	// copy the pod as this method modifies the pod's node selectors to force it to meet topology constraints
+	pod = pod.DeepCopy()
+
+	var possibleNodes []*Node
+	for _, node := range s.nodes {
+		if err := node.Compatible(ctx, pod); err == nil {
+			possibleNodes = append(possibleNodes, node)
```

Comment:
If we're convinced that we want to go with this approach, can we map to the kube-scheduler's idioms of `filter` and `score`

Reply:
We can, but I was trying to minimize changes in this PR since it was already 1k+.